### ISSUE TITLE
MM3 energy is non-smooth: JAX grad returns wrong subgradients at converged metal-TS geometries

### ISSUE THREAD

**Issue by ericchansen** [ericchansen/q2mm#284]
# MM3 energy is non-smooth: JAX `grad` returns wrong subgradients at converged metal-TS geometries

## TL;DR

The MM3 energy function has non-smooth points (most likely `arccos` near 0/π, or vdW switching) where `jax.grad` returns a **wrong gradient** — disagreeing with finite-difference by orders of magnitude.  Optimizations that land near such points pick descent directions that aren't actually downhill on the real surface.  This is an exploitable correctness bug that biases optimization results, not just numerical noise.

A separate but related issue is that `ObjectiveFunction(x)` is **noisy** on GPU for these systems (0.6%–5% per-call spread).  That noise is a measurement uncertainty we should live with via statistical methods (median-of-N + CI) — it isn't a bug to chase.

This issue tracks the gradient correctness bug; the noise is documented below for context.

---

## §1 — Confirmed bug: JAX-grad disagrees with FD at converged geometries

At the initial QM geometry, `jax.grad` matches finite-difference to ~1e-6.  After the geometry minimizer runs to convergence on the rh-conjugate metal-TS, `jax.grad` and FD disagree by ~700 in norm:

```
Initial QM geom: JAX grad norm = 1602.08, FD grad norm = 1602.08, max |JAX-FD| = 1.06e-6
BFGS converged:  JAX grad norm = 1671.66, FD grad norm =  713.49, max |JAX-FD| = 7.10e+2
```

The 700-norm disagreement is too large to be floating-point noise — it's a wrong subgradient.  The optimizer follows the bad gradient into apparent stationary points that aren't actually stationary, and any reported "improvement" in that regime is suspect.

### Reproduction

```python
import jax, jax.numpy as jnp, numpy as np
from scipy.optimize import minimize as scipy_min
from q2mm.diagnostics.systems import load_system
from q2mm.backends.mm.jax_engine import JaxEngine

sys_data = load_system('rh-conjugate', engine=JaxEngine())
ff = sys_data.forcefield
mol = sys_data.molecules[0]
eng = JaxEngine()
handle = eng._get_handle(mol, ff)
params, coords = eng._params_and_coords(handle, ff)
energy_fn = handle._energy_fn
grad_fn = jax.jit(jax.grad(lambda c, p: energy_fn(p, c.reshape(-1, 3)), argnums=0))

# Use BFGS (Python, deterministic) to reach the converged geometry
def fun(x): return float(energy_fn(params, jnp.asarray(x).reshape(-1, 3)))
def jac(x): return np.asarray(grad_fn(jnp.asarray(x), params))
x0 = np.asarray(coords.flatten())
r = scipy_min(fun, x0, jac=jac, method='BFGS', options={'maxiter': 3000, 'gtol': 1e-8})
xf = r.x

# Compare JAX grad vs FD at the converged point
g_jax = jac(xf)
g_fd = np.zeros_like(xf)
eps = 1e-5
for i in range(len(xf)):
    xp = xf.copy(); xp[i] += eps
    xm = xf.copy(); xm[i] -= eps
    g_fd[i] = (fun(xp) - fun(xm)) / (2 * eps)
print(f'JAX grad norm: {np.linalg.norm(g_jax):.3e}')
print(f'FD  grad norm: {np.linalg.norm(g_fd):.3e}')
print(f'max |JAX-FD|:  {np.max(np.abs(g_jax - g_fd)):.3e}')
```

### Conjectured causes (in order of plausibility)

1. **`arccos` in bond-angle term near 0 or π.** The derivative of `arccos(x)` is `-1/sqrt(1-x²)` which blows up at the endpoints.  JAX returns ±∞ or NaN at exact 0/π and gets a spuriously large value just shy of the endpoints, while FD averages across the kink.
2. **vdW switching function** (cutoff smoothing) where the second derivative is piecewise.  Same kink-averaging story.
3. **`jnp.where`-based piecewise terms** anywhere in the energy expression.  JAX's VJP for `where` is well-known to leak NaN gradients from the unused branch when the chosen branch is non-finite.

### Why "wrong gradient", not "noise"

A noisy gradient would still average to the true gradient over many samples; it would have zero bias.  This is a structural bias: every time the optimizer lands near a particular geometry, JAX-grad reports the same wrong value, while FD reports a different (smaller, in our case) value.  Reproduces deterministically on CPU.

---

## §2 — Known noise (not a bug, just a measurement uncertainty)

For full context, `ObjectiveFunction(x)` is non-deterministic on GPU for metal-TS systems due to scipy `L-BFGS-B`'s Fortran implementation carrying hidden state across calls.  Per-system 5-call IQR/median:

| System | Spread | Note |
|---|---|---|
| ch3f | 0.00 % | Deterministic — single mol, no metal |
| rh-enamide | 0.58 % | |
| pd-allyl | 0.65 % | |
| rh-conjugate | 2.10 % | |
| heck-relay | 4.90 % | Worst — also has a JaxLoss surrogate breakdown |

### Reproduction

```python
from q2mm.diagnostics.systems import load_system
from q2mm.backends.mm.jax_engine import JaxEngine
from q2mm.optimizers.objective import ObjectiveFunction

sys_data = load_system('rh-conjugate', engine=JaxEngine())
ff = sys_data.forcefield
x0 = ff.get_param_vector().copy()
obj = ObjectiveFunction(ff, JaxEngine(), sys_data.molecules, sys_data.reference)
for i in range(3):
    print(f'call {i}:', obj(x0))
# call 0: 6284858.567284835
# call 1: 6430076.3873276785
# call 2: 6492788.971720077
```

CPU is deterministic; the energy function and gradient function (when evaluated directly) are deterministic on GPU; the non-determinism enters in scipy's `_lbfgsb.f` line-search bookkeeping when the gradient evaluations happen to differ in the low bits (TF32 etc.).

### Why this isn't worth fixing directly

Per-call noise of 0.6–5 % is well within what statistical methods can handle.  Median-of-N evaluations plus a confidence interval gives a publishable improvement claim that's robust against this kind of noise.  Engineering effort is better spent on §1 (the actual bug) than on either chasing GPU-determinism flags (slow, may not even fully work) or rewriting `engine.minimize` to use a deterministic-but-slower optimizer.

A companion PR adds `--n-evals` + median+CI reporting to `scripts/regenerate_convergence_results.py` — see whichever PR lands the change.  Once that ships, every per-system page in the docs can replace "within noise floor" with a proper median ± CI verdict.

---

## §3 — Path forward (this issue's action items)

Two-step plan, both targeting §1:

### Step A — Bisect which MM3 term is non-smooth

For each candidate term (`arccos` angle, vdW switching, any `jnp.where`), construct a minimal repro:

1. Single bond/angle, parameters that put the geometry exactly at the non-smooth point.
2. Compare `jax.grad` vs FD at that geometry.

This pins down which term is responsible.  Without that, any "fix" is shooting in the dark.

### Step B — Fix the offending term

Two options, depending on what step A finds:

- **Preferred — smooth the term.**  E.g. replace `arccos(x)` with a regularized form near the endpoints, or replace a hard cutoff in vdW with a smooth tapering function.  Fixes the bug for all callers; no surprises for anyone differentiating through MM3 in the future.
- **Fallback — custom JAX VJP rule.**  Register a `jax.custom_vjp` that returns the FD-correct gradient at the non-smooth point.  Less elegant, but doesn't change the forward pass.

### What this issue is NOT about

- Fixing scipy L-BFGS-B's hidden state — it's noise, see §2.
- Rewriting `JaxEngine.minimize` — that may end up being the implementation strategy for §3.B, but it's the means, not the goal.

## References

- q2mm#283 — the e2e optimization PR where the noise was first quantified.
- q2mm-data#6 — the matching data PR with the 5-call noise measurements per system.

**Comment by ericchansen** [comment]
Closing — both halves of this issue are now resolved:

**§1 (gradient correctness bug)** — fixed by #288 (merged). `arccos(clip(cos θ))` in the MM3 angle term was replaced with a `jax.custom_vjp`-wrapped `atan2(|a×b|, a·b)` formulation that returns analytic gradients with `sin θ` floored at 1e-12. Regression test in `test/test_mm3_jax.py::test_near_collinear_gradient_matches_fd`.

**§2 (per-call noise)** — addressed earlier by #286 (median-of-N + 95% CI) and #287. The convergence harness now reports statistical significance instead of point estimates.

**Headline impact:** two systems went from "no improvement" to significant gains once the gradient bug was fixed:

| System | Pre-fix | Post-fix |
|---|---:|---:|
| rh-conjugate | NOT SIG | **18.00% ± 4.17%** |
| heck-relay | NOT SIG | **52.82% ± 1.54%** |

Validated artifacts: ericchansen/q2mm-data#10 (merged).
